applications/AdjointFluidApplication/custom_schemes/adjoint_steady_scheme.h

```diff
@@ -206,10 +225,27 @@ class AdjointSteadyScheme : public Scheme<TSparseSpace, TDenseSpace>
     {
         KRATOS_TRY
 
-        for (auto it = rDofSet.begin(); it != rDofSet.end(); ++it)
-            if (it->IsFree() == true)
-                it->GetSolutionStepValue() +=
-                    TSparseSpace::GetValue(rDx, it->EquationId());
+        Communicator& rComm = rModelPart.GetCommunicator();
+
+        if (rComm.TotalProcesses() == 1)
+        {
+            for (auto it = rDofSet.begin(); it != rDofSet.end(); ++it)
+                if (it->IsFree() == true)
+                    it->GetSolutionStepValue() +=
+                        TSparseSpace::GetValue(rDx, it->EquationId());
+        }
+        else
+        {
+            for (auto it = rDofSet.begin(); it != rDofSet.end(); ++it)
+                if (it->GetSolutionStepValue(PARTITION_INDEX) == rComm.MyPID())
+                    if (it->IsFree() == true)
+                        it->GetSolutionStepValue() +=
+                            TSparseSpace::GetValue(rDx, it->EquationId());
+
+            // todo: add a function Communicator::SynchronizeDofVariables() to
+            // reduce communication here.
+            rComm.SynchronizeNodalSolutionStepsData();
```

Comment:
I'm not sure about this. It looks like the trilinos builder and solver uses a map where only the dofs owned by the partition are local. We can discuss it this afternoon.